include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1182,37 +1215,35 @@ __parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPoli
 
-                    __dpl_sycl::__atomic_ref<_AtomicType, sycl::access::address_space::global_space> __found(
-                        *__dpl_sycl::__get_accessor_ptr(__temp_acc));
-                    __dpl_sycl::__atomic_ref<_AtomicType, sycl::access::address_space::local_space> __found_local(
-                        *__dpl_sycl::__get_accessor_ptr(__temp_local));
+                    // 1. Set initial value to local found state
+                    _AtomicType __found_local = __init_value;
 
-                    // 1. Set initial value to local atomic
-                    if (__local_idx == 0)
-                        __found_local.store(__init_value);
-                    __dpl_sycl::__group_barrier(__item_id);
-
-                    // 2. Find any element that satisfies pred and set local atomic value to global atomic
+                    // 2. Find any element that satisfies pred
                     __pred(__item_id, __n_iter, __wgroup_size, __found_local, __brick_tag, __rngs...);
-                    __dpl_sycl::__group_barrier(__item_id);
 
-                    // Set local atomic value to global atomic
-                    if (__local_idx == 0)
+                    // 3. Reduce over group: find __dpl_sycl::__minimum (for the __parallel_find_forward_tag),
+                    // find __dpl_sycl::__maximum (for the __parallel_find_backward_tag)
+                    // or calculate the sum with __dpl_sycl::__plus (for the __parallel_or_tag)
+                    // inside all our group items
+                    __found_local = __dpl_sycl::__reduce_over_group(__item_id.get_group(), __found_local,
+                                                                    typename _BrickTag::_LocalResultsReduceOp{});
+
+                    // Set local found state value value to global atomic
+                    if (__local_idx == 0 && __found_local != __init_value)
                     {
-                        const auto __found_local_state = __found_local.load();
-                        if (__found_local_state != __init_value)
-                        {
-                            // Update global (for all groups) atomic state with the found index
-                            _BrickTag::__save_state_to_atomic(__found, __found_local_state);
-                        }
+                        __dpl_sycl::__atomic_ref<_AtomicType, sycl::access::address_space::global_space> __found(
+                            *__dpl_sycl::__get_accessor_ptr(__result_sycl_buf_acc));
+
+                        // Update global (for all groups) atomic state with the found index
+                        _BrickTag::__save_state_to_atomic(__found, __found_local);
                     }
                 });
         });
         //The end of the scope  -  a point of synchronization (on temporary sycl buffer destruction)
     }
 
     if constexpr (__or_tag_check)
-        return __result;
+        return __result != __init_value;
```

Comment:
Is this change really needed? Wouldn't `__result` be 0 anyways if no match is found?

Reply:
We should return the result of `bool` type in this case, so this expression looks like more correct then implicit conversion of some numeric type to `bool` type, I believe.

Reply:
Also this expression looks more clear from my point of view.

Reply:
I'm fine with your change. I just think there is no semantical difference. We could switch to `bool` in general. I don't expect much performance benefit from this though since this is only a single value in registers in `__early_exit_find_or` and we need an atomic type in the global update anyways.

Reply:
This `bool` returning value is not my change: it has place before.

Reply:
Resolved?

Reply:
My comment is not about the return type but the internal `_BrickTag::_AtomicType`. Anyways, there is no change required from my point of view.